File: frontend/src/api/teacher.ts
/**
 * The AI teacher: streaming lecture, grounded/agentic Q&A, and offline library
 * chat. Each is a `text/event-stream` POST decoded through the shared readSSE.
 */

import { readSSE } from './sse'
import type { GraphNode, GraphEdge } from './graph'

/**
 * One beat of a lecture: a paragraph of narration bound to the graph nodes
 * it's about, so they light up in sync as the beat is revealed.
 */
export interface Beat {
  /** 3–6 word signpost shown above the paragraph. */
  heading: string
  /** The narration paragraph itself. */
  text: string
  /** Ids of the graph nodes to highlight while this beat is on screen. */
  node_ids: string[]
}

/**
 * A trimmed node the teacher needs — the visible graph is its grounding
 * scope, so only the fields that feed the prompt are sent.
 */
export interface TeacherNode {
  id: string
  title: string
  year: number | null
  citation_count?: number | null
  authors?: string | null
  tldr?: string | null
  abstract?: string | null
  rels: string[]
}

/**
 * What story the lecture tells: the field's history, the seed paper's
 * intuition, or a conceptual bridge from the seed to a target paper.
 */
export type LectureMode = 'history' | 'intuition' | 'bridge'

/**
 * A backward-in-time hop the history lecture took before narrating
 * (Phase 3e): how many foundational ancestors it pulled in and the oldest
 * year it reached.
 */
export interface LectureTrace {
  hop: number
  found: number
  oldest: number | null
  /** The hop hit an S2 error / rate limit rather than empty results. */
  error?: boolean
}

/** Callbacks for {@link streamLecture}. Optional ones may be omitted. */
export interface LectureHandlers {
  /** A new beat arrived — append it to the lecture panel. */
  onBeat: (beat: Beat) => void
  /** History-mode backward hop progress (precedes the beats). */
  onTrace?: (t: LectureTrace) => void
  /** Ancestors pulled in by the backward walk, to merge into the graph. */
  onNodes?: (d: Discovery) => void
  onDone?: () => void
  onError?: (message: string) => void
  /** Abort to cancel the stream (e.g. when the user closes the panel). */
  signal?: AbortSignal
}

/**
 * Stream a lecture over the visible graph. Beats arrive one at a time. In
 * history mode, trace + nodes events (the backward walk to the field's
 * roots) precede them.
 *
 * @param body The seed, the visible nodes, the lecture mode, and (bridge
 *             mode only) the target paper.
 * @param h    Event handlers; see {@link LectureHandlers}.
 */
export async function streamLecture(
  body: { seed: { title: string; id?: string }; nodes: TeacherNode[]; mode: LectureMode; target?: { title: string } },
  h: LectureHandlers,
): Promise<void> {
  const res = await fetch('/api/lecture', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
    signal: h.signal,
  })
  await readSSE(res, (event, data) => {
    if (event === 'beat') h.onBeat(data as Beat)
    else if (event === 'trace') h.onTrace?.(data as LectureTrace)
    else if (event === 'nodes') h.onNodes?.(data as Discovery)
    else if (event === 'done') h.onDone?.()
    else if (event === 'error') h.onError?.((data as { error: string }).error)
  })
}

/**
 * A step the agent took — reading a paper, expanding the graph to one not
 * yet shown, or searching for off-graph papers. Surfaced live in the chat as
 * the agent works.
 */
export interface TraceEvent {
  action: 'read' | 'expand' | 'search' | 'search_sources'
  ok: boolean
  title?: string | null
  index?: number
  /** 'summary' | 'full' — read_paper. */
  detail?: string
  /** 'references' | 'citations' | 'similar' — expand_node. */
  relation?: string
  /** New papers discovered / passages found. */
  found?: number
  /** Free-text query — search_papers / search_sources. */
  query?: string
  /** Year filter — search_papers. */
  year_from?: number | null
  year_to?: number | null
}

/**
 * New papers (+ the edges connecting them) the agent pulled in via
 * expand_node / search_papers, to be merged into the live graph.
 */
export interface Discovery {
  nodes: GraphNode[]
  edges: GraphEdge[]
}

/** Callbacks for {@link streamAsk}. Optional ones may be omitted. */
export interface AskHandlers {
  /** A chunk of answer prose arrived — append it to the streaming bubble. */
  onToken: (text: string) => void
  /** The final citation list: ids of the papers the answer drew from. */
  onCited: (nodeIds: string[]) => void
  /** An agent step (read/expand/search) to render as a trace chip. */
  onTrace?: (t: TraceEvent) => void
  /** Papers the agent discovered, to merge into the live graph. */
  onNodes?: (d: Discovery) => void
  /** Drop streamed preamble that turned out to precede a tool call. */
  onDiscard?: () => void
  onDone?: () => void
  onError?: (message: string) => void
  /** Abort to cancel the stream mid-answer. */
  signal?: AbortSignal
}

/**
 * Stream a grounded answer: agent trace steps, prose tokens, then the nodes
 * it drew from. (Non-agentic backends just emit tokens + cited.)
 *
 * @param body The question, a session id for follow-up context, the seed,
 *             the visible nodes (the grounding scope), and an optional
 *             source_id scoping the agent's library search to one uploaded
 *             source (agentic backend only).
 * @param h    Event handlers; see {@link AskHandlers}.
 */
export async function streamAsk(
  body: {
    question: string
    session_id: string
    seed: { title: string; id?: string }
    nodes: TeacherNode[]
    source_id?: string
  },
  h: AskHandlers,
): Promise<void> {
  const res = await fetch('/api/ask', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
    signal: h.signal,
  })
  await readSSE(res, (event, data) => {
    if (event === 'token') h.onToken((data as { text: string }).text)
    else if (event === 'trace') h.onTrace?.(data as TraceEvent)
    else if (event === 'nodes') h.onNodes?.(data as Discovery)
    else if (event === 'discard') h.onDiscard?.()
    else if (event === 'cited') h.onCited((data as { node_ids: string[] }).node_ids)
    else if (event === 'done') h.onDone?.()
    else if (event === 'error') h.onError?.((data as { error: string }).error)
  })
}

/**
 * Offline library chat (Phase 3d): the one trace it emits — which passages
 * the retrieval pulled, and from which sources — shown above the grounded
 * answer.
 */
export interface RetrieveEvent {
  found: number
  sources: string[]
}

/** Callbacks for {@link streamAskSources}. Optional ones may be omitted. */
export interface AskSourcesHandlers {
  /** A chunk of answer prose arrived. */
  onToken: (text: string) => void
  /** The retrieval summary (emitted once, before any prose). */
  onRetrieve?: (r: RetrieveEvent) => void
  onDone?: () => void
  onError?: (message: string) => void
  signal?: AbortSignal
}

/**
 * Stream an answer grounded purely in the user's local library — no graph.
 * A single retrieve event, then prose tokens.
 *
 * @param body The question, a session id for follow-up context, and an
 *             optional source_id to scope retrieval to one source.
 * @param h    Event handlers; see {@link AskSourcesHandlers}.
 */
export async function streamAskSources(
  body: { question: string; session_id: string; source_id?: string },
  h: AskSourcesHandlers,
): Promise<void> {
  const res = await fetch('/api/ask_sources', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
    signal: h.signal,
  })
  await readSSE(res, (event, data) => {
    if (event === 'token') h.onToken((data as { text: string }).text)
    else if (event === 'trace') h.onRetrieve?.(data as RetrieveEvent)
    else if (event === 'done') h.onDone?.()
    else if (event === 'error') h.onError?.((data as { error: string }).error)
  })
}

File: backend/arxiv_digest/routes/teacher.py
"""AI-teacher routes: a streamed lecture over the visible graph, grounded/agentic
Q&A, and the offline library chat. All three stream Server-Sent Events.

POST /api/lecture      -> streamed AI lecture over the visible graph
POST /api/ask          -> streamed grounded Q&A over the visible graph
POST /api/ask_sources  -> streamed chat answered purely from the local library
"""

from __future__ import annotations

import json
from typing import Iterator

from flask import Blueprint, Response, current_app, jsonify, request
from flask.typing import ResponseReturnValue

from .. import config
from .. import teacher as teacher_service
from ..library import sources

bp = Blueprint("teacher", __name__)

# Session-scoped Q&A history, kept in memory (cleared on restart — fine for v1).
# Maps a client-generated session id -> [{role, content}, ...].
_QA_SESSIONS: dict[str, list[dict]] = {}
# Separate history store for the offline library chat (Phase 3d) — same shape,
# kept apart so a graph Q&A and a library chat don't cross-contaminate context.
_SOURCES_SESSIONS: dict[str, list[dict]] = {}


def _opt_source_ids(payload: dict) -> "list[str] | None":
    """Parse the optional ``source_ids`` scope from a request body.

    Args:
        payload: The parsed JSON body.

    Returns:
        The library source ids to scope the answer to. ``None`` when the key is
        absent or malformed — no scope, so the whole library is searched. A
        **present** ``source_ids`` array yields exactly its string entries,
        including an **empty list** — an explicit "no sources selected" that
        searches nothing rather than everything.
    """
    raw = payload.get("source_ids")
    if not isinstance(raw, list):
        return None
    return [s for s in raw if isinstance(s, str) and s]


def _sse(event: str, data: object) -> str:
    """Format one Server-Sent Event frame.

    Args:
        event: The event name (``beat``, ``token``, ``trace``, …).
        data: A JSON-serializable payload.

    Returns:
        The wire-format frame: ``event:`` and ``data:`` lines terminated by a
        blank line.

    Raises:
        TypeError: When ``data`` isn't JSON-serializable.
    """
    return f"event: {event}\ndata: {json.dumps(data)}\n\n"


def _sse_response(generator) -> Response:
    """Wrap a generator of SSE frames as a streaming response.

    ``X-Accel-Buffering: no`` keeps nginx (if ever put in front) from
    buffering the stream; ``Cache-Control: no-cache`` stops intermediaries
    caching partial output.

    Args:
        generator: An iterator yielding SSE frame strings (from ``_sse``).

    Returns:
        A ``text/event-stream`` Flask Response streaming the frames.
    """
    return Response(
        generator,
        mimetype="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )


@bp.post("/api/lecture")
def api_lecture() -> ResponseReturnValue:
    """Stream a lecture over the visible graph as SSE ``beat`` events.

    In ``history`` mode we first walk backward through references (Phase 3e),
    emitting ``trace`` + ``nodes`` events, so the story can start at the
    field's roots; the discovered ancestors join the node set the lecture
    narrates over.

    Body:
        ``{seed: {title,...}, nodes: [visible node objects], mode:
        history|intuition|bridge, target?: {title,...}}``.

    Returns:
        An SSE stream — each ``beat`` event carries ``{heading, text,
        node_ids}`` so the frontend can reveal it and light up nodes, ending
        with ``done`` (or an ``error`` event; failures inside the stream are
        also logged). HTTP 400 when ``nodes`` is missing/empty.
    """
    payload = request.get_json(silent=True) or {}
    nodes = payload.get("nodes")
    if not isinstance(nodes, list) or not nodes:
        return jsonify({"error": "nodes must be a non-empty list"}), 400
    seed = payload.get("seed") or {}
    mode = payload.get("mode") or "history"
    target = payload.get("target")

    def gen() -> Iterator[str]:
        """Yield the lecture's SSE frames (backfill first in history mode)."""
        try:
            enriched = nodes
            if mode == "history" and seed.get("id"):
                for kind, data in teacher_service.history_backfill(seed, nodes):
                    if kind == "nodes":
                        assert isinstance(data, dict)  # "nodes" events carry {nodes, ...}
                        enriched = enriched + data["nodes"]
                        yield _sse("nodes", data)
                    elif kind == "trace":
                        yield _sse("trace", data)
            for beat in teacher_service.lecture_beats(seed, enriched, mode=mode, target=target):
                yield _sse("beat", beat)
            yield _sse("done", {})
        except Exception as exc:  # surface to the panel AND log the traceback
            current_app.logger.exception("lecture failed")
            yield _sse("error", {"error": str(exc)})

    return _sse_response(gen())


@bp.post("/api/ask")
def api_ask() -> ResponseReturnValue:
    """Answer a question grounded in the visible graph, streamed as SSE.

    Runs the agentic Q&A (reads papers via tool use) when the API backend is
    available; otherwise the non-agentic grounded answer (e.g. under the
    claude CLI backend). Conversation history is keyed by ``session_id`` so
    follow-ups keep context; a turn is persisted only on success, capped to
    the recent window.

    Body:
        ``{question, session_id, seed, nodes, source_ids?}`` — ``source_ids``
        scopes the agent's library search to a subset of uploaded sources
        (agentic backend only).

    Returns:
        An SSE stream: with the agentic backend, ``trace`` events (tool
        steps), ``nodes`` events (``{nodes, edges}`` discoveries) as expansion
        finds papers not yet on the graph, and ``figure`` events (a paper's
        figure the agent attached to its answer); always ``token`` events
        (prose), a final ``cited`` event (``{node_ids}``), then ``done`` (or
        ``error``). HTTP 400 when the question or nodes are missing.
    """
    payload = request.get_json(silent=True) or {}
    question = (payload.get("question") or "").strip()
    if not question:
        return jsonify({"error": "question is required"}), 400
    nodes = payload.get("nodes")
    if not isinstance(nodes, list) or not nodes:
        return jsonify({"error": "nodes must be a non-empty list"}), 400
    seed = payload.get("seed") or {}
    session_id = payload.get("session_id") or ""
    source_ids = _opt_source_ids(payload)  # scope the teacher to a subset of sources
    history = _QA_SESSIONS.get(session_id, []) if session_id else []

    # Agentic Q&A (reads papers via tool use) when the API backend is available;
    # otherwise the non-agentic grounded answer (e.g. the claude CLI backend).
    # The source scope only bears on the agentic path's library search; the
    # non-agentic grounded answer is graph-only and ignores it.
    source = (
        teacher_service.answer_agentic(question, seed, nodes, history, source_ids)
        if teacher_service.agentic_available()
        else teacher_service.answer_stream(question, seed, nodes, history)
    )

    def gen() -> Iterator[str]:
        """Yield the answer's SSE frames, persisting the turn on success."""
        answer_parts: list[str] = []
        try:
            for kind, data in source:
                if kind == "token":
                    assert isinstance(data, str)  # "token" events carry prose
                    answer_parts.append(data)
                    yield _sse("token", {"text": data})
                elif kind == "trace":
                    yield _sse("trace", data)
                elif kind == "nodes":
                    yield _sse("nodes", data)
                elif kind == "figure":
                    yield _sse("figure", data)
                elif kind == "discard":
                    # Streamed preamble turned out to precede a tool call — drop it.
                    answer_parts.clear()
                    yield _sse("discard", {})
                elif kind == "cited":
                    yield _sse("cited", {"node_ids": data})
            yield _sse("done", {})
        except Exception as exc:
            current_app.logger.exception("ask failed")
            yield _sse("error", {"error": str(exc)})
            return
        # Persist the turn only on success, capped to the recent window.
        if session_id:
            convo = _QA_SESSIONS.setdefault(session_id, [])
            convo.append({"role": "user", "content": question})
            convo.append({"role": "assistant", "content": "".join(answer_parts).strip()})
            keep = config.TEACHER_HISTORY_TURNS * 2
            if len(convo) > keep:
                del convo[:-keep]

    return _sse_response(gen())


@bp.post("/api/ask_sources")
def api_ask_sources() -> ResponseReturnValue:
    """Answer a question purely from the user's local library, streamed as SSE.

    The offline library chat — no graph required. History is keyed by
    ``session_id`` (a separate store from the graph Q&A) so follow-ups keep
    context without cross-contaminating the two chats.

    Body:
        ``{question, session_id, source_ids?}`` — ``source_ids`` scopes
        retrieval to a subset of sources.

    Returns:
        An SSE stream: one ``trace`` event (the retrieved passages), then
        ``token`` prose, then ``done`` (or ``error``). HTTP 400 when the
        question is missing or the local library is unavailable
        (embeddings/sqlite-vec didn't load).
    """
    payload = request.get_json(silent=True) or {}
    question = (payload.get("question") or "").strip()
    if not question:
        return jsonify({"error": "question is required"}), 400
    if not sources.available():
        return jsonify({"error": "Your local library is unavailable (embeddings/sqlite-vec didn't load)."}), 400

    session_id = payload.get("session_id") or ""
    source_ids = _opt_source_ids(payload)  # scope to a subset of sources (optional)
    history = _SOURCES_SESSIONS.get(session_id, []) if session_id else []

    def gen() -> Iterator[str]:
        """Yield the library answer's SSE frames, persisting the turn on success."""
        answer_parts: list[str] = []
        try:
            for kind, data in teacher_service.answer_from_sources(question, history, source_ids):
                if kind == "token":
                    assert isinstance(data, str)  # "token" events carry prose
                    answer_parts.append(data)
                    yield _sse("token", {"text": data})
                elif kind == "trace":
                    yield _sse("trace", data)
            yield _sse("done", {})
        except Exception as exc:
            current_app.logger.exception("ask_sources failed")
            yield _sse("error", {"error": str(exc)})
            return
        if session_id:
            convo = _SOURCES_SESSIONS.setdefault(session_id, [])
            convo.append({"role": "user", "content": question})
            convo.append({"role": "assistant", "content": "".join(answer_parts).strip()})
            keep = config.TEACHER_HISTORY_TURNS * 2
            if len(convo) > keep:
                del convo[:-keep]

    return _sse_response(gen())
